### PR TITLE
BUG: pd.ExcelFile closes stream on destruction

### LINKED ISSUE #31467: pd.ExcelFile closes stream on destruction in pandas 1.0.0
#### Code Sample, a copy-pastable example if possible

```python3

import pandas as pd
print(pd.__version__)
a = open('some_file.xlsx', 'rb')
x = pd.ExcelFile(a)
del x
print(a.read())
```

#### Problem description

Above script behaves in different way in pandas 0.25.3 and 1.0.0:

```bash
$ python3 t1.py
0.25.3
b''
$ sudo pip3 install pandas --upgrade --quiet
$ python3 t1.py
1.0.0
Traceback (most recent call last):
  File "t1.py", line 6, in <module>
    print(a.read())
ValueError: read of closed file
```

It seems that stream is closed when ExcelFile is destroyed - and I don't see why.

#### Expected Output

I'd expect either notice in release notes, or the same output in 0.25.3 and 1.0.0.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.8.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.0.0-1028-gcp
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : C.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.0
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 9.0.1
setuptools       : 39.0.1
Cython           : None
pytest           : 4.3.0
hypothesis       : None
sphinx           : 1.8.5
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : 0.999999999
pymysql          : None
psycopg2         : 2.8.3 (dt dec pq3 ext lo64)
jinja2           : 2.10.1
IPython          : None
pandas_datareader: None
bs4              : 4.8.0
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : 2.5.14
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 4.3.0
pyxlsb           : None
s3fs             : None
scipy            : 1.2.0
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by roberthdevries**
- [x] closes #31467
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/excel/test_readers.py [hunk 1]
```diff
@@ -1020,10 +1031,10 @@ def test_excel_read_buffer(self, engine, read_ext):
         tm.assert_frame_equal(expected, actual)
 
     def test_reader_closes_file(self, engine, read_ext):
-        f = open("test1" + read_ext, "rb")
-        with pd.ExcelFile(f) as xlsx:
-            # parses okay
-            pd.read_excel(xlsx, "Sheet1", index_col=0, engine=engine)
+        with open("test1" + read_ext, "rb") as f:
```
Comment (WillAyd) [inline pandas/tests/io/excel/test_readers.py]: Can you revert this? Not totally equivalent

File: pandas/tests/io/excel/test_openpyxl.py [hunk 2]
```diff
@@ -114,7 +114,7 @@ def test_to_excel_with_openpyxl_engine(ext, tmpdir):
     df2 = DataFrame({"B": np.linspace(1, 20, 10)})
     df = pd.concat([df1, df2], axis=1)
     styled = df.style.applymap(
-        lambda val: "color: %s" % "red" if val < 0 else "black"
+        lambda val: "color: %s" % ("red" if val < 0 else "black")
```
Comment (WillAyd) [inline pandas/tests/io/excel/test_openpyxl.py]: Can you revert this? OK to clean up but better as a separate PR

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/io/excel/test_readers.py (hunk 1 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_readers.py]: It is indeed not equivalent, but that is on purpose. The test is faulty. First you open a file and fail to close it, then you check for a file leak. This demonstrates in fact the issue we are having. Ownership of the file handle should be managed at the level it is created, unless you explicitly transfer ownership of the file handle to `ExcelFile`, but IMHO that is not what you would expect (and the reason for the ticket)

File: pandas/tests/io/excel/test_openpyxl.py (hunk 2 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_openpyxl.py]: It is a fix in a test, not really a code fix. But I can move it into a separate PR if you prefer that.

File: doc/source/whatsnew/v1.1.0.rst [hunk 3]
```diff
@@ -302,6 +302,7 @@ I/O
   timestamps with ``version="2.0"`` (:issue:`31652`).
 - Bug in :meth:`read_csv` was raising `TypeError` when `sep=None` was used in combination with `comment` keyword (:issue:`31396`)
 - Bug in :class:`HDFStore` that caused it to set to ``int64`` the dtype of a ``datetime64`` column when reading a DataFrame in Python 3 from fixed format written in Python 2 (:issue:`31750`)
+- Bug in :class:`ExcelFile` where the stream passed into the function was closed by the destructor. (:issue:`31467`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.1.0.rst]: Is this a regression? If so, it should be in 1.0.2.rst.

File: pandas/io/excel/_base.py [hunk 4]
```diff
@@ -366,6 +366,10 @@ def _workbook_class(self):
     def load_workbook(self, filepath_or_buffer):
         pass
 
+    @abc.abstractmethod
```
Comment (TomAugspurger) [inline pandas/io/excel/_base.py]: Mmm does this have to be abstract? That may be too large a change for a point release.

Comment (TomAugspurger) [reply pandas/io/excel/_base.py]: And it seems like several children just pass, so can we just make this non-abstract?

File: pandas/tests/io/excel/test_openpyxl.py (hunk 2 above)
Comment (TomAugspurger) [reply pandas/tests/io/excel/test_openpyxl.py]: I'd prefer to revert this if it's unrelated, just to minimize the chance of a conflict when backporting.

**Review by TomAugspurger — COMMENTED** [review]
It seems like there's some disagreement over https://github.com/pandas-dev/pandas/pull/32544/files#r389428225. Can you resolve that?

Keep in mind that if we're backporting this to 1.0.2 we want the changes to be as minimal as necessary.

File: pandas/tests/io/excel/test_openpyxl.py (hunk 2 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_openpyxl.py]: I'll make a separate PR for this and remove it in this one.

Comment (roberthdevries) [reply pandas/tests/io/excel/test_openpyxl.py]: See  #32563

File: pandas/io/excel/_base.py (hunk 4 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py]: OK, will fix

File: pandas/tests/io/excel/test_openpyxl.py (hunk 2 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_openpyxl.py]: Moved to PR #32563

File: doc/source/whatsnew/v1.1.0.rst (hunk 3 above)
Comment (roberthdevries) [reply doc/source/whatsnew/v1.1.0.rst]: Moved to v1.0.2.rst
Regarding the disagreement: undoing the change makes the test fail on the file leak check.
The only way this test will not leak a file, is by undoing this fix. So either we keep this bug as is and leave the test unchanged, or we fix the bug and fix the broken test.
Please note that it is the test code itself that is leaking the file descriptor, not the `ExcelFile` class

File: pandas/tests/io/excel/test_readers.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/tests/io/excel/test_readers.py]: @WillAyd do you know what the original intent of this test was? The blame history just points to a test reorganization.

Seems like the disagreement is around whether `ExcelFile(obj).__exit__` should close the file when given an open file object.

@roberthdevries can you check the behavior here on pandas 0.25.3 and 1.0.1? And the behavior on this branch is to not close the file when provided an open file?

I don't think this changed test is adding much value. It's just testing the behavior of `open` now, which we can probably rely on being correct :)

Comment (WillAyd) [reply pandas/tests/io/excel/test_readers.py]: Yea I think this test strays a little bit from the OP, where a straight ExcelFile object is closing the handle. In this case the context manager is closing it, which maybe is OK. I can't speak to original history of this.

So @roberthdevries can you maybe summarize what you are proposing? Something like:

 1. ExcelFile Context Manager + path: should manage open/close lifecycle
 2. ExcelFile Context Manager + handle: should not manage open/close lifecycle
 3. ExcelFile + path: should manage open, not close
 4. ExcelFile + handle: should not manage open/close lifecycle

? 

I think point 2 above is a little ambiguous; if there's a precedent for it elsewhere would be interested

Comment (roberthdevries) [reply pandas/tests/io/excel/test_readers.py]: I think the basic philosophy should be that a resource should be released at the same level where it was created. This is one of the main use cases of context managers.
So whenever a resource is released at **another** level, you violate this rule. (hence this ticket)
The original `ExcelFile` code closed a file it got from the outside, and this specific test relied on this (erroneous IMHO) behavior.
So a nested context manager should work as expected, in the sense that the outside context manager releases the outer resource and the inner context manager releases the inner resource.

So:
```python
with open("foo.xlsx") as f:
    with ExcelFile(f) as xlsx:
        # do stuff
```
In this specific example the `ExcelFile` context manager should **not** close `f`, as that would be the responsibility of the `open` context manager.

In this example:
```python
f = open("foo.xlsx")
with ExcelFile(f) as xlsx:
    # do stuff
```
We have a file leak caused by the programmer forgetting to close the file handle. This is a programmer error, and IMHO not something that needs to be fixed.
However, whenever `f` goes out of scope, the `__del__` method will be called which will probably close the file anyway.

For use cases where `ExcelFile` is used in combination with path names the context manager case implies closing the file.
In this case:
```python
xlsx = ExcelFile(filename)
```
I would propose to have a `__del__` method which will close the file and the other `ExcelFile` resources.

So I agree with you on case 1 and 4.
I also agree with case 2, there is no ambiguity for me. The `ExcelFile` context manager only should manage its own resources, not the file resource.
For case 3 there should be a nuance in the sense that the `__del__` method should close the resource.
For case 4, the `ExcelFile` should not touch the open/close lifecyle of the file handle, but the `ExcelFile` resource itself should be closed again in the `__del__` method.

File: pandas/tests/io/excel/test_readers.py [hunk 5]
```diff
@@ -629,6 +629,17 @@ def test_read_from_py_localpath(self, read_ext):
...
         tm.assert_frame_equal(expected, actual)
 
+    @td.check_file_leaks
+    def test_close_from_py_localpath(self, read_ext):
+
+        # GH31467
+        str_path = os.path.join("test1" + read_ext)
+        with open(str_path, "rb") as f:
+            x = pd.read_excel(f, "Sheet1", index_col=0)
+            del x
```
Comment (WillAyd) [inline pandas/tests/io/excel/test_readers.py]: Is there a way to construct this test without the `del` call? This test might not be doing anything, since `del` doesn't really mean `__del__` gets called

File: pandas/tests/io/excel/test_readers.py (hunk 1 above)
Comment (WillAyd) [reply pandas/tests/io/excel/test_readers.py]: Thanks for the insights. So agree with what you are saying but not sure that is clearly expressed in the implementation. Shouldn't we alternately just within the ExcelFile construct or `__enter__` keep track of whether or not the resource lifecycle is owned or borrowed? As is I don't think that's ever very clear

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/io/excel/test_openpyxl.py [hunk 6]
```diff
@@ -114,7 +114,7 @@ def test_to_excel_with_openpyxl_engine(ext, tmpdir):
     df2 = DataFrame({"B": np.linspace(1, 20, 10)})
     df = pd.concat([df1, df2], axis=1)
     styled = df.style.applymap(
-        lambda val: "color: %s" % ("red" if val < 0 else "black")
+        lambda val: "color: %s" % "red" if val < 0 else "black"
```
Comment (jbrockmendel) [inline pandas/tests/io/excel/test_openpyxl.py]: is this related to the rest of the PR?

File: pandas/tests/io/excel/test_readers.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/tests/io/excel/test_readers.py]: > Shouldn't we alternately just within the ExcelFile constructor

That seems most sensible to me. Do we have a standard method for checking if something is a file-like object vs. a path?

File: pandas/io/excel/_base.py [hunk 7]
```diff
@@ -366,6 +366,9 @@ def _workbook_class(self):
     def load_workbook(self, filepath_or_buffer):
         pass
 
+    def close(self):
```
Comment (datapythonista) [inline pandas/io/excel/_base.py]: Would it make sense to make this an abstract method, and make child classes be explicit on how the close is performed (including when no close is performed)?

File: pandas/io/excel/_base.py [hunk 8]
```diff
@@ -895,14 +898,7 @@ def sheet_names(self):
...
     def close(self):
         """close io if necessary"""
-        if self.engine == "openpyxl":
-            # https://stackoverflow.com/questions/31416842/
-            #  openpyxl-does-not-close-excel-workbook-in-read-only-mode
-            wb = self.book
-            wb._archive.close()
-
-        if hasattr(self.io, "close"):
-            self.io.close()
```
Comment (datapythonista) [inline pandas/io/excel/_base.py]: Is this intentionally removed? Wondering if there is any engine that needs this...

**Review by datapythonista — APPROVED** [review]
thanks for the fix @roberthdevries

**Comment by datapythonista** [comment]
You've got a conflict, if you can merge master and fix it please.

File: pandas/io/excel/_base.py (hunk 7 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py]: In an earlier version of this PR, this was an abstract method, but the powers that be preferred this.

File: pandas/io/excel/_base.py (hunk 8 above)
Comment (roberthdevries) [reply pandas/io/excel/_base.py]: This is the bug fix

**Comment by roberthdevries** [comment]
Rebased to master

File: pandas/tests/io/excel/test_openpyxl.py (hunk 6 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_openpyxl.py]: Removed

File: pandas/io/excel/_base.py (hunk 7 above)
Comment (TomAugspurger) [reply pandas/io/excel/_base.py]: We can't make a new abstract method in a bugfix release. Any child not implementing it would fail when the class is instantiated.

File: pandas/tests/io/excel/test_readers.py (hunk 5 above)
Comment (roberthdevries) [reply pandas/tests/io/excel/test_readers.py]: This is what was in the original test (just indented), but can probably be removed.

On another note, the test title: `test_reader_closes_file` seems to imply that the intent of the test is to verify that `read_excel`  closes the file (or maybe `ExcelFile`). Which seems very counter intuitive to me.

File: pandas/tests/io/excel/test_readers.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/tests/io/excel/test_readers.py]: It ends up not being an issue here since ExcelFile doesn't open any files, just the engine.

**Review by TomAugspurger — APPROVED** [review]
I merged master to fix CI and the merge conflict. Should be good to go.

**Comment by TomAugspurger** [comment]
Thanks @roberthdevries.
